vllm/v1/core/sched/scheduler.py

```diff
@@ -1468,44 +1502,66 @@ def _update_requests_with_invalid_blocks(
         return affected_req_ids, total_affected_tokens
 
     def _handle_invalid_blocks(self, invalid_block_ids: set[int]) -> set[str]:
-        total_requests_to_reschedule = 0
-        total_tokens_to_reschedule = 0
+        """
+        Handle requests affected by invalid KV cache blocks.
 
-        # --- Handle async KV loads (WAITING_FOR_REMOTE_KVS) ---
+        Returns:
+            Set of affected request IDs to skip in update_from_output main loop.
+        """
+        should_fail = (
+            self.vllm_config.kv_transfer_config is not None
+            and self.vllm_config.kv_transfer_config.kv_load_failure_policy == "fail"
+        )
+
+        # Handle async KV loads (WAITING_FOR_REMOTE_KVS)
         async_load_reqs = (
             req
             for req in self.waiting
             if req.status == RequestStatus.WAITING_FOR_REMOTE_KVS
         )
-        async_affected_req_ids, num_tokens_to_reschedule = (
+        async_failed_req_ids, num_failed_tokens = (
             self._update_requests_with_invalid_blocks(
                 async_load_reqs, invalid_block_ids
             )
         )
 
-        total_requests_to_reschedule += len(async_affected_req_ids)
-        total_tokens_to_reschedule += num_tokens_to_reschedule
-
-        # Mark requests with async KV load failures; they will be rescheduled
-        # once loading completes.
-        self.failed_recving_kv_req_ids |= async_affected_req_ids
+        total_failed_requests = len(async_failed_req_ids)
+        total_failed_tokens = num_failed_tokens
 
-        # --- Handle sync KV loads (running requests) ---
-        sync_affected_req_ids, num_tokens_to_reschedule = (
+        sync_failed_req_ids, num_failed_tokens = (
             self._update_requests_with_invalid_blocks(self.running, invalid_block_ids)
         )
 
-        total_requests_to_reschedule += len(sync_affected_req_ids)
-        total_tokens_to_reschedule += num_tokens_to_reschedule
-
-        if total_requests_to_reschedule:
-            logger.warning(
-                "Recovered from KV load failure: "
-                "%d request(s) rescheduled (%d tokens affected).",
-                total_requests_to_reschedule,
-                total_tokens_to_reschedule,
+        total_failed_requests += len(sync_failed_req_ids)
+        total_failed_tokens += num_failed_tokens
+
+        if not total_failed_requests:
+            return set()
+
+        # evict invalid blocks from the prefix cache to prevent future
+        # requests from reusing corrupted data. this must happen regardless
+        # of retry policy to prevent cache pollution race conditions.
+        self.kv_cache_manager.evict_blocks(invalid_block_ids)
```

Comment:
Async loading blocks won't be cached at this pint, does that matter?

Reply:
found a cheap fix for this in f479682a11612d79f20314339318e609e5ede698, before eviction we can simply check if the blocks have hashes in the block table, if they don't they can be ignored.